simplifyTypes raised: Option

I obtain an internal compiler error (`simplifyTypes raised: Option`) when I try to compile MLKit with MLton version 20201002 under macOS (Catalina). Here is how it occurs:

````
$ mlton
MLton 20201002
$ uname -a
Darwin Martins-MBP 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64
$ git clone https://github.com/melsman/mlkit.git
$ cd mlkit
$ git checkout v4.5.1
$ ./autobuild
$ ./configure --with-compiler='mlton -verbose 2'
$ make mlkit
````

Here is the last part of the output from the last command:

````
$ make mlkit
...
mlton -verbose 2 -output mlkit Compiler/native64.mlb
MLton 20201002 starting
   Compile SML starting
      frontend starting
         parseAndElaborate starting
         parseAndElaborate finished in 4.17 + 2.42 (37% GC)
         deadCode starting
         deadCode finished in 0.07 + 0.00 (0% GC)
         defunctorize starting
         defunctorize finished in 0.74 + 0.39 (34% GC)
      frontend finished in 4.98 + 2.81 (36% GC)
      xmlSimplify starting
         xmlTypeCheck starting
         xmlTypeCheck finished in 0.37 + 0.00 (0% GC)
         xmlShrink starting
         xmlShrink finished in 0.48 + 0.00 (0% GC)
         xmlSimplifyTypes starting
         xmlSimplifyTypes finished in 0.15 + 0.00 (0% GC)
         xmlTypeCheck starting
         xmlTypeCheck finished in 0.31 + 0.00 (0% GC)
      xmlSimplify finished in 1.32 + 0.00 (0% GC)
      monomorphise starting
      monomorphise finished in 1.56 + 0.46 (23% GC)
      sxmlSimplify starting
         sxmlTypeCheck starting
         sxmlTypeCheck finished in 0.68 + 0.00 (0% GC)
         sxmlShrink1 starting
         sxmlShrink1 finished in 1.35 + 0.00 (0% GC)
         implementSuffix starting
         implementSuffix finished in 0.06 + 0.44 (88% GC)
         sxmlShrink2 starting
         sxmlShrink2 finished in 0.93 + 0.00 (0% GC)
         implementExceptions starting
         implementExceptions finished in 0.13 + 0.00 (0% GC)
         sxmlShrink3 starting
         sxmlShrink3 finished in 1.09 + 0.00 (0% GC)
         polyvariance starting
            duplicate1 starting
            duplicate1 finished in 1.71 + 0.38 (18% GC)
            duplicate2 starting
            duplicate2 finished in 1.30 + 0.00 (0% GC)
         polyvariance finished in 3.01 + 0.38 (11% GC)
         sxmlShrink4 starting
         sxmlShrink4 finished in 0.71 + 0.39 (35% GC)
         sxmlTypeCheck starting
         sxmlTypeCheck finished in 0.47 + 0.00 (0% GC)
      sxmlSimplify finished in 8.45 + 1.21 (12% GC)
      closureConvert starting
         flow analysis starting
         flow analysis finished in 0.71 + 0.00 (0% GC)
         free variables starting
         free variables finished in 0.24 + 0.00 (0% GC)
         globalize starting
         globalize finished in 0.19 + 0.00 (0% GC)
         convert starting
         convert finished in 4.36 + 7.67 (64% GC)
      closureConvert finished in 5.82 + 7.67 (57% GC)
      ssaSimplify starting
         ssaTypeCheck starting
         ssaTypeCheck finished in 3.19 + 0.00 (0% GC)
         removeUnused1 starting
         removeUnused1 finished in 4.24 + 2.24 (35% GC)
         introduceLoops1 starting
         introduceLoops1 finished in 0.08 + 0.00 (0% GC)
         loopInvariant1 starting
         loopInvariant1 finished in 1.32 + 1.25 (49% GC)
         inlineLeaf1 starting
         inlineLeaf1 finished in 1.33 + 0.00 (0% GC)
         inlineLeaf2 starting
         inlineLeaf2 finished in 1.33 + 1.87 (58% GC)
         contify1 starting
         contify1 finished in 1.29 + 0.00 (0% GC)
         localFlatten1 starting
         localFlatten1 finished in 1.46 + 1.19 (45% GC)
         constantPropagation starting
         constantPropagation finished in 4.23 + 6.19 (59% GC)
         duplicateGlobals1 skipped
         splitTypes1 starting
         splitTypes1 finished in 11.58 + 3.92 (25% GC)
         useless starting
         useless finished in 9.17 + 10.51 (53% GC)
         loopUnroll1 skipped
         removeUnused2 starting
         removeUnused2 finished in 2.26 + 7.37 (77% GC)
         duplicateGlobals2 starting
         duplicateGlobals2 finished in 0.14 + 0.52 (79% GC)
         splitTypes2 starting
         splitTypes2 finished in 11.93 + 2.01 (14% GC)
         simplifyTypes starting
            simplifyTypes raised: Option
         simplifyTypes raised in 0.16 + 0.00 (0% GC)
      ssaSimplify raised in 53.70 + 37.06 (41% GC)
   Compile SML raised in 75.84 + 49.20 (39% GC)
MLton 20201002 raised in 75.84 + 49.20 (39% GC)
make[1]: *** [mlkit_kit] Error 1
make: *** [mlkit] Error 2
bash-3.2$ 
````

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -1,5 +1,11 @@
 = CHANGELOG
 
+== Version YYYYMMDD
+
+* 2020-10-22
+  ** Fix bug in `SimplifyTypes` SSA optimization pass.  Thanks
+  to Martin Elsman for the bug report.
+
 == Version 20201002
 
 Here are the changes from version 20200817 to version 20201002.
@@ -12,7 +18,8 @@ Here are the changes from version 20200817 to version 20201002.
 === Details
 
 * 2020-09-20
-  ** Fix bug in handling of weak objects during mark-compact GC.
+  ** Fix bug in handling of weak objects during mark-compact GC.  Thanks to
+  Bernard Berthomieu for the bug report.
 
 == Version 20200817
 
@@ -123,10 +130,10 @@ Here are the changes from version 20180206 to version 20200722.
   non-printable.
 
 * 2020-02-14
-  ** Fixed bug in `SimplifyTypes` SSA optimization pass.
+  ** Fix bug in `SimplifyTypes` SSA optimization pass.
 
 * 2020-01-22
-  ** Added expert `-pi-style {default|npi|pic|pie}` and
+  ** Add expert `-pi-style {default|npi|pic|pie}` and
   `-native-pic {false|true}` options, which can be used to override a
   target-determined default.  See
   https://github.com/MLton/mlton/pull/365 for details.

--- a/mlton/Makefile
+++ b/mlton/Makefile
@@ -75,7 +75,7 @@ $(MLTON_OUTPUT): $(SOURCES)
 		-target $(TARGET) -output $(MLTON_OUTPUT)			\
 		$(MLTON_MLB)
 
-ifeq ($(shell (cat control/version_sml.src; echo $(MLTON_NAME) $(MLTON_VERSION); if [ -e control/version.sml ]; then cat control/version.sml; fi) | $(SHA1DGST)),$(shell if [ -e control/version_sml.chk ]; then cat control/version_sml.chk; fi))
+ifeq ($(shell (cat control/version_sml.src; echo '$(MLTON_NAME)' '$(MLTON_VERSION)'; if [ -e control/version.sml ]; then cat control/version.sml; fi) | $(SHA1DGST)),$(shell if [ -e control/version_sml.chk ]; then cat control/version_sml.chk; fi))
 control/version.sml: control/version_sml.src
 	touch control/version.sml
 else
@@ -86,7 +86,7 @@ control/version.sml: control/version_sml.src
 		-e "s/MLTON_VERSION/$(MLTON_VERSION)/" \
 		< control/version_sml.src \
 		> control/version.sml
-	(cat control/version_sml.src; echo $(MLTON_NAME) $(MLTON_VERSION); cat control/version.sml) | $(SHA1DGST) > control/version_sml.chk
+	(cat control/version_sml.src; echo '$(MLTON_NAME)' '$(MLTON_VERSION)'; cat control/version.sml) | $(SHA1DGST) > control/version_sml.chk
 endif
 
 front-end/%.lex.sml: front-end/%.lex

--- a/mlton/ssa/simplify-types.fun
+++ b/mlton/ssa/simplify-types.fun
@@ -346,7 +346,11 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
                              Cardinality.layout (tyconCardinality tycon)]);
                Vector.foreach
                (cons, fn {con, ...} =>
-                (display (seq [str "cardinality of ",
+                (display (seq [str "rep of ",
+                               Con.layout con,
+                               str " = ",
+                               ConRep.layout (conRep con)]);
+                 display (seq [str "cardinality of ",
                                Con.layout con,
                                str " = ",
                                Cardinality.layout (conCardinality con)])))))
@@ -355,10 +359,10 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
          (setTyconNumCons (tycon, 1)
           ; setTyconReplacement (tycon, SOME (Type.tuple args))
           ; setConRep (con, ConRep.Transparent))
-      (* "unary" is datatypes with one constructor whose rhs contains an
-       * array (or vector) type.
-       * For datatypes with one variant not containing an array type, eliminate
-       * the datatype.
+      (* "unary" are datatypes with one constructor
+       * whose rhs contains an array (or vector) type.
+       * For datatypes with one variant not containing an array type,
+       * eliminate the datatype.
        *)
       fun containsArrayOrVector (ty: Type.t): bool =
          let
@@ -436,6 +440,30 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
                   :: accum
           else (transparent (tycon, con, args)
                 ; accum))
+      (* diagnostic *)
+      val _ =
+         Control.diagnostics
+         (fn display =>
+          let
+             open Layout
+          in
+             Vector.foreach
+             (origDatatypes, fn Datatype.T {tycon, cons} =>
+              (display (seq [str "num cons of ",
+                             Tycon.layout tycon,
+                             str " = ",
+                             Int.layout (tyconNumCons tycon)]);
+               display (seq [str "replacement of ",
+                             Tycon.layout tycon,
+                             str " = ",
+                             Option.layout Type.layout (tyconReplacement tycon)]);
+               Vector.foreach
+               (cons, fn {con, ...} =>
+                display (seq [str "rep of ",
+                              Con.layout con,
+                              str " = ",
+                              ConRep.layout (conRep con)]))))
+          end)
 
       fun makeSimplifyTypeFns simplifyType =
          let

--- a/mlton/ssa/simplify-types.fun
+++ b/mlton/ssa/simplify-types.fun
@@ -247,7 +247,7 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
                          Prim.MLton_bogus =>
                             (case Type.dest (Vector.sub (targs, 0)) of
                                 Type.Datatype tycon =>
-                                   Cardinality.makeMany (tyconCardinality tycon)
+                                   Cardinality.makeOne (tyconCardinality tycon)
                               | _ => ())
                        | _ => ())
                 | _ => ()
@@ -298,7 +298,6 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
                | Word _ => Cardinality.many
            end))
       (* Remove useless constructors from datatypes.
-       * Remove datatypes which have no cons.
        * Lower-bound cardinality of cons by product of arguments.
        * Lower-bound cardinality of tycons by sum of cons.
        *)
@@ -308,7 +307,7 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
       val typeIsVoid = fn t => Type.equals (t, typeVoid)
       val origDatatypes = datatypes
       val datatypes =
-         Vector.keepAllMap
+         Vector.map
          (datatypes, fn Datatype.T {tycon, cons} =>
           let
              val cons = Vector.keepAll (cons, conIsUseful o #con)
@@ -325,11 +324,7 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
                   (Vector.map (args, typeCardinality)),
                   conCardinality con))
           in
-             if Vector.isEmpty cons
-                then (setTyconNumCons (tycon, 0)
-                      ; setTyconReplacement (tycon, SOME typeVoid)
-                      ; NONE)
-             else SOME (Datatype.T {tycon = tycon, cons = cons})
+             Datatype.T {tycon = tycon, cons = cons}
           end)
       (* diagnostic *)
       val _ =
@@ -392,7 +387,9 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
           in
              case Vector.length cons of
                 0 => (setTyconNumCons (tycon, 0)
-                      ; setTyconReplacement (tycon, SOME typeVoid)
+                      ; if Cardinality.isZero (tyconCardinality tycon)
+                           then setTyconReplacement (tycon, SOME typeVoid)
+                           else setTyconReplacement (tycon, SOME Type.unit)
                       ; (datatypes, unary))
               | 1 =>
                    let

--- a/mlton/ssa/simplify-types.fun
+++ b/mlton/ssa/simplify-types.fun
@@ -439,7 +439,15 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
 
       fun makeSimplifyTypeFns simplifyType =
          let
+            val simplifyType =
+               Trace.trace
+               ("SimplifyTypes.simplifyType", Type.layout, Type.layout)
+               simplifyType
             fun simplifyTypes ts = Vector.map (ts, simplifyType)
+            val simplifyTypes =
+               Trace.trace ("SimplifyTypes.simplifyTypes",
+                            Vector.layout Type.layout, Vector.layout Type.layout)
+               simplifyTypes
             fun simplifyUsefulTypesOpt ts =
                Exn.withEscape
                (fn escape =>
@@ -454,9 +462,20 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
                              then NONE
                           else SOME t
                        end)))
+            val simplifyUsefulTypesOpt =
+               Trace.trace ("SimplifyTypes.simplifyUsefulTypesOpt",
+                            Vector.layout Type.layout,
+                            Option.layout (Vector.layout Type.layout))
+               simplifyUsefulTypesOpt
             val simplifyUsefulTypes = valOf o simplifyUsefulTypesOpt
+            val simplifyUsefulTypes =
+               Trace.trace ("SimplifyTypes.simplifyUsefulTypes",
+                            Vector.layout Type.layout,
+                            Vector.layout Type.layout)
+               simplifyUsefulTypes
          in
-            {simplifyTypes = simplifyTypes,
+            {simplifyType = simplifyType,
+             simplifyTypes = simplifyTypes,
              simplifyUsefulTypes = simplifyUsefulTypes,
              simplifyUsefulTypesOpt = simplifyUsefulTypesOpt}
          end
@@ -466,7 +485,7 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
           Property.initRec
           (fn (t, simplifyType) =>
            let
-              val {simplifyUsefulTypesOpt, ...} =
+              val {simplifyType, simplifyUsefulTypesOpt, ...} =
                  makeSimplifyTypeFns simplifyType
               fun doitPtr (mk, t) =
                  let
@@ -499,10 +518,7 @@ fun transform (Program.T {datatypes, globals, functions, main}) =
                | Weak t => doitPtr (weak, t)
                | _ => t
            end))
-      val simplifyType =
-         Trace.trace ("SimplifyTypes.simplifyType", Type.layout, Type.layout)
-         simplifyType
-      val {simplifyTypes, simplifyUsefulTypesOpt, simplifyUsefulTypes, ...} =
+      val {simplifyType, simplifyTypes, simplifyUsefulTypesOpt, simplifyUsefulTypes, ...} =
          makeSimplifyTypeFns simplifyType
       val typeIsUseful = not o Type.isUnit o simplifyType
       (* Simplify constructor argument types. *)

